crates/provider/src/native.rs

```diff
@@ -0,0 +1,307 @@
+use std::{
+    self,
+    error::Error,
+    fmt::Debug,
+    process::{Command, Output},
+};
+
+use async_trait::async_trait;
+use serde::Serialize;
+
+use crate::shared::{
+    constants::{DEFAULT_DATA_DIR, DEFAULT_REMOTE_DIR, LOCALHOST},
+    provider::Provider,
+    types::{NamespaceDef, NamespaceMetadata, RunCommandOptions, RunCommandResponse},
+};
+
+pub trait FileSystem {
+    fn create_dir(&mut self, path: impl Into<String>) -> Result<(), Box<dyn Error>>;
+    fn write(
+        &mut self,
+        path: impl Into<String>,
+        content: impl Into<String>,
+    ) -> Result<(), Box<dyn Error>>;
+}
+
+#[derive(Debug, Serialize)]
+struct FilesystemInMemory {}
+
+#[derive(Debug, Serialize, Clone, PartialEq)]
+struct NativeProvider<T: FileSystem + Debug> {
+    /// Namespace of the client
+    namespace:             String,
+    /// path where configuration relies
+    config_path:           String,
+    // variable that shows if debug is activated
+    debug:                 bool,
+    /// the timeout for the client to exit
+    timeout:               u32,
+    /// command sent to client
+    command:               String,
+    /// temporary directory
+    tmp_dir:               String,
+    pod_monitor_available: bool,
+    local_magic_file_path: String,
+    remote_dir:            String,
+    data_dir:              String,
+    filesystem:            T,
+}
+
+impl<T: FileSystem + Debug> NativeProvider<T> {
+    pub fn new(
+        namespace: impl Into<String>,
+        config_path: impl Into<String>,
+        tmp_dir: impl Into<String>,
+        filesystem: T,
+    ) -> Self {
+        let tmp_dir = tmp_dir.into();
+
+        Self {
+            namespace: namespace.into(),
+            config_path: config_path.into(),
+            debug: true,
+            timeout: 60, // seconds
+            local_magic_file_path: format!("{}/finished.txt", &tmp_dir),
+            remote_dir: format!("{}{}", &tmp_dir, DEFAULT_REMOTE_DIR),
+            data_dir: format!("{}{}", &tmp_dir, DEFAULT_DATA_DIR),
+            command: "bash".into(),
+            tmp_dir,
+            pod_monitor_available: false,
+            filesystem,
+        }
+    }
+}
+
+#[async_trait]
+impl<T: FileSystem + Debug> Provider for NativeProvider<T> {
+    fn create_namespace(&mut self) -> Result<(), Box<dyn Error>> {
+        let name_space_def = NamespaceDef {
+            api_version: "v1".into(),
+            kind:        "Namespace".into(),
+            metadata:    NamespaceMetadata {
+                name:   format!("{}", &self.namespace),
+                labels: None,
+            },
+        };
+
+        let file_path = format!("{}/{}", &self.tmp_dir, "namespace");
+        let content = serde_json::to_string(&name_space_def)?;
+
+        self.filesystem.write(file_path, content)?;
+        self.filesystem.create_dir(&self.remote_dir)?;
```

Comment:
Why do we create a dir after writing the file ? @pepoviola any idea ?

Reply:
This can be refactored, from this `ts` code

```ts
  async createNamespace(): Promise<void> {
    const namespaceDef = {
      apiVersion: "v1",
      kind: "Namespace",
      metadata: {
        name: this.namespace,
      },
    };

    writeLocalJsonFile(this.tmpDir, "namespace", namespaceDef);
    // Native provider don't have the `namespace` isolation.
    // but we create the `remoteDir` to place files
    await makeDir(this.remoteDir, true);
    return;
  }
  ```
  
  For native we should only do
  
  ```rust
            fn create_namespace(&mut self) -> Result<(), Box<dyn Error>> { 
              self.filesystem.create_dir(&self.remote_dir)?;
              Ok(())
            }
```